preprocess/foldfilter_main.cc

```diff
@@ -0,0 +1,320 @@
+	for (size_t i = 0; i < delimiters.size(); ++i)
+		if (character == delimiters[i])
+			return i;
+
+	return not_found;
+}
+
+pair<deque<StringPiece>,deque<string>> wrap_lines(StringPiece const &line, wrap_options const &options) {
+	deque<StringPiece> out_lines;
+
+	deque<string> out_delimiters;
+
+	// Current byte position
+	int32_t pos = 0;
+
+	// Length of line in bytes
+	int32_t length = line.size();
+	
+	// Byte position of last cut-off point
+	int32_t pos_last_cut = 0;
+
+	// For each delimiter the byte position of its last occurrence
+	vector<int32_t> pos_delimiters(options.delimiters.size(), 0);
+
+	// Position of the first delimiter we encountered up to pos. Reset
+	// to pos + next char if it's not a delimiter.
+	int32_t pos_first_delimiter = 0;
+
+	while (pos < length) {
+		UChar32 character;
+
+		U8_NEXT(line.data(), pos, length, character);
+		
+		if (character < 0)
+			throw utf8::NotUTF8Exception(line);
+
+		size_t delimiter_idx = find_delimiter(options.delimiters, character);
+
+		if (delimiter_idx != not_found) {
+			// Store pos_first_delimiter instead of pos because when we have
+			// consecutive delimiters we want to chop em all off, even when
+			// our ideal delimiter is somewhere in the middle.
+			pos_delimiters[delimiter_idx] = pos_first_delimiter;
+		} else {
+			// Maybe the next char is a delimiter? pos is pointing to the next
+			// one right now, U8_NEXT incremented it.
+			pos_first_delimiter = pos;
+		}
+
+		// Do we need to introduce a break? If not, move to next character
+		if (pos - pos_last_cut < static_cast<int32_t>(options.column_width))
+			continue;
+
+		// Last resort if we didn't break on a delimiter: just chop where we are
+		int32_t pos_cut = pos;
+
+		// Find a more ideal break point by looking back for a delimiter
+		for (int32_t const &pos_delimiter : pos_delimiters) {
+			if (pos_delimiter > pos_last_cut) {
+				pos_cut = pos_delimiter;
+				break;
+			}
+		}
+
+		// Assume we cut without delimiters (i.e. the last resort scenario)
+		int32_t pos_cut_end = pos_cut;
+
+		// Peek ahead to were after the cut we encounter our first not-a-delimiter
+		// because that's the point were we resume.
+		for (int32_t pos_next = pos_cut_end; pos_cut_end < length; pos_cut_end = pos_next) {
+			// When we're not skipping delimiters, don't send more bytes than
+			// column_width in total a single line, even though we try to keep
+			// the delimiters together.
+			if (options.keep_delimiters && pos_cut_end - pos_last_cut >= static_cast<int32_t>(options.column_width))
+				break;
+
+			U8_NEXT(line.data(), pos_next, length, character);
+
+			if (character < 0)
+				throw utf8::NotUTF8Exception(line);
+
+			// First character after pos_cut is probably a delimiter, unless
+			// we did a hard stop in the middle of a word, and we're not keeping
+			// the delimiters.
+			if (find_delimiter(options.delimiters, character) == not_found)
+				break;
+		}
+
+		if (options.keep_delimiters) {
```

Comment:
Is this the wrong way around?  if keep_delimiters then `out_delimiters.emplace_back("");`

Reply:
Keep in the sense that the delimiters go to the child process. The glue added by foldfilter itself (out_delimiters) will then be just an empty string.

Reply:
Oh.  Maybe better naming?

Reply:
Renamed it to keep_delimiters_in_lines.